spec/systems/users_spec.rb

```diff
@@ -0,0 +1,162 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe 'Signup', type: :system do
+  context 'validates screen' do
+    it 'displays signup screen should show correct field' do
+      visit signup_path
+
+      expect(page).to have_content('Signup')
+
+      within 'form' do
+        expect(page).to have_field('user[username]')
+        expect(page).to have_field('user[password]')
+        expect(page).to have_field('user[password_confirmation]')
+      end
+    end
+
+    it 'displays signup screen show correct label' do
+      visit signup_path
+
+      expect(page).to have_content('Signup')
+
+      within 'form' do
+        expect(page).to have_field('user[username]')
+        expect(page).to have_field('user[password]')
+        expect(page).to have_field('user[password_confirmation]')
+      end
+    end
+  end
+
+  context 'given valid data' do
+    it 'redirects to index page and show flash message' do
+      visit signup_path
+
+      within 'form' do
+        fill_in('user[username]', with: 'nimblehq')
+        fill_in('user[password]', with: 'password')
+        fill_in('user[password_confirmation]', with: 'password')
+
+        click_button('Signup')
+      end
+      expect(current_path).to eql(users_path)
+      expect(page).to have_content('Account was successfully created')
+    end
+  end
+
+  context 'given invalid data' do
+    it 'displays password not match error when password and password confirmation are not match' do
+      visit signup_path
+
+      within 'form' do
+        fill_in('user[username]', with: 'nimblehq')
+        fill_in('user[password]', with: 'password')
+        fill_in('user[password_confirmation]', with: 'drowssap')
+
+        click_button('Signup')
+
+        expect(page).to have_selector('#errorExplanation')
+        within '#errorExplanation' do
+          expect(page).to have_content('Password not match')
```

Comment:
In general, we don't test the sad path on UI test because UI is slower than unit test so we only test the happy path on UI test.

Also for this test, it's already covered in User Model unit test.

Reply:
In this case, I want to test if errors are display on signup page when signup failed.

Reply:
yeah in general we don't test that case :D we trust the Active record error 😄

Reply:
You can take a look on this really good book also @gutakk https://books.thoughtbot.com/assets/testing-rails.pdf

Reply:
Fixed in https://github.com/gutakk/ruby-google-scraper/pull/4/commits/ab300567f17b0438ad058fa58d6d3f5363c4268d